fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.registry;
+
+import com.google.common.collect.ImmutableMap;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.util.Block2ObjectMap;
+import net.fabricmc.fabric.impl.content.registry.SpreadableBlockRegistryImpl;
+
+/**
+ * Registries of Blocks to BlockStates, defining the spreadable BlockState to replace
+ * a bare Block with when a particular type of spreadable Block spreads to the bare black.
+ *
+ * <p>For example, to get the registry for Mycelium block spread and then register
+ * a modded block pair where Mycelium spreading to MY_DIRT generates MY_MYCELIUM:
+ *
+ * <pre>{@code
+ * SpreadableBlockRegistry.getOrCreateInstance(SpreadableBlockRegistry.MYCELIUM)
+ *        .add(MyModBlocks.MY_DIRT, MyModBlocks.MY_MYCELIUM.getDefaultState());
+ * }</pre>
+ */
+@ApiStatus.NonExtendable
+public interface SpreadableBlockRegistry extends Block2ObjectMap<BlockState> {
+	/**
+	 * Registry ID for Minecraft Grass Block type spreadable blocks.
+	 */
+	Identifier GRASS = new Identifier("grass");
+	/**
+	 * Registry ID for Minecraft Mycelium Block type spreadable blocks.
+	 */
+	Identifier MYCELIUM = new Identifier("mycelium");
+
+	/**
+	 * Get the registry for a given registry ID, or create a new registry for the ID
+	 * if none currently exists.
+	 *
+	 * @param type The registry type Identifier for the desired spreadable block registry
+	 * @return The SpreadableBlockRegistry for the given ID
+	 */
+	static SpreadableBlockRegistry getOrCreateInstance(Identifier type) {
+		return SpreadableBlockRegistryImpl.getOrCreateInstance(type);
+	}
+
+	/**
+	 * Gets an immutable copy of the spreadable block types map, which maps the
+	 * canonical or "primary" block of the type to the registry for the type.
```

Comment:
Doesn't `SpreadableBlockRegistry.add` map *all* blocks of the type to registry, not just a "primary" block?

https://github.com/FabricMC/fabric/blob/725e336a53e8714456674d2250107f2479cbfc4e/fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/SpreadableBlockRegistryImpl.java#L97-L102

Reply:
Yes, `SpreadableBlockRegistry.add` is called to add all the spreadable block transitions to the associated registries.  The types map is basically a map of registry key (which is the primary spreadable block) to registry.  The map is used by the vanilla mixin to implement vanilla-to-mod spread.  Spread from a modded block must be implemented in the block.

You could argue the API implementation is not complete without a `FabricSpreadableBlock` type thing with a default implementation of the mod->any logic.  I wouldn't argue against it; this PR probably should be rewritten anyway, now that MixinExtras is available in the API.  When I have some time in January, I'd like to take another pass at trying to implement some of the more complicated requests (such as allowing mods to set the new block state).

Reply:
Right. I was just wondering if it's really the "primary" spreadable block if it applies to *all* spreadable blocks using the registry.